contracts/cw721-stake/src/contract.rs

```diff
@@ -0,0 +1,527 @@
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapper: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if info.sender != config.nft_address {
+        return Err(ContractError::InvalidToken {
+            received: info.sender,
+            expected: config.nft_address,
+        });
+    }
+
+    let sender = deps.api.addr_validate(&wrapper.sender)?;
+
+    STAKED_NFTS_PER_OWNER.update(
+        deps.storage,
+        sender.clone(),
+        env.block.height,
+        |nft_collection| -> StdResult<IndexSet<String>> {
+            let mut updated_nft_collection = nft_collection.unwrap_or_default();
+            updated_nft_collection.insert(wrapper.token_id.clone());
+            Ok(updated_nft_collection)
+        },
+    )?;
+
+    TOTAL_STAKED_NFTS.update(
+        deps.storage,
+        env.block.height,
+        |total_staked| -> StdResult<_> {
+            total_staked
+                .unwrap()
+                .checked_add(Uint128::new(1))
+                .map_err(StdError::overflow)
+        },
+    )?;
+
+    let hook_msgs = stake_hook_msgs(deps.storage, sender.clone(), wrapper.token_id.clone())?;
+    Ok(Response::default()
+        .add_submessages(hook_msgs)
+        .add_attribute("action", "stake")
+        .add_attribute("from", sender)
+        .add_attribute("token_id", wrapper.token_id))
+}
+
+pub fn execute_unstake(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    token_ids: Vec<String>,
+) -> Result<Response, ContractError> {
+    if token_ids.is_empty() {
+        return Err(ContractError::ZeroUnstake {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+
+    let resulting_collection = STAKED_NFTS_PER_OWNER.update(
+        deps.storage,
+        info.sender.clone(),
+        env.block.height,
+        |nft_collection| -> Result<IndexSet<String>, ContractError> {
+            if let Some(mut nft_collection) = nft_collection {
+                // Some benchmarking suggests this is actually the
+                // fastest way to remove a list of items from a
+                // HashSet.
+                //
+                // Alternatives include, `drain().filter()` and
+                // `difference(&to_remove.into_iter().collect()).cloned().collect()`. Difference
+                // here, suprisingly, being ~2x the speed of drain and
+                // filter. Remove in a loop clocks in at ~2x the speed
+                // of difference.
+                for token_id in token_ids.iter() {
+                    // This will implicitly check for duplicates in
+                    // the input vector as removing twice will fail
+                    // the second time around.
+                    let was_present = nft_collection.remove(token_id);
+                    if !was_present {
+                        // Can't unstake that which you do not own.
+                        return Err(ContractError::NotStaked {});
```

Comment:
if there's a token ID which they do not own in there, should we still unstake the valid ones, or should we fail the whole operation as we do now?  

I think it's probably fine as is but just thought I'd call it out

Reply:
Good point. I'd probably prefer failing the whole operation here. If you're attempting to unstake a NFT you don't own something is likely amis. Think it also fits into the general "transaction either completely succeeds or completely fails" model that blockchains tend to use.